File: wiki/scratch/parse_crafting.js
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

const dataDir = path.resolve('../../crafting data');
const outFile = path.resolve('../Module/Crafting_SLASH_Data.md');

function toTitleCase(s) {
    if (!s) return "";
    return s.split('_').map(w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join(' ');
}

function parseSlotItem(itemStr) {
    itemStr = itemStr.trim();
    if (itemStr.startsWith('v AIR') || !itemStr) return null;

    let parts = itemStr.split(' ');
    let amount = "1";
    let name = "Unknown";

    if (itemStr.startsWith('v ')) {
        let material = parts[1];
        name = toTitleCase(material);
        for (let p of parts) {
            if (p.includes('..')) amount = p.split('.')[0];
        }
    } else if (itemStr.startsWith('m ')) {
        let idPart;
        let idx = parts.indexOf('-');
        if (idx !== -1) {
            idPart = parts[idx - 1];
        } else {
            idPart = parts[2];
        }
        name = toTitleCase(idPart);
        for (let p of parts) {
            if (p.includes('..')) amount = p.split('.')[0];
        }
    }

    if (!isNaN(parseInt(amount)) && parseInt(amount) > 1) {
        return name + "," + amount;
    }
    return name;
}

let luaTable = {};

let files = fs.readdirSync(dataDir).filter(f => f.endsWith('.yml'));

for (let file of files) {
    let filePath = path.join(dataDir, file);
    try {
        let content = fs.readFileSync(filePath, 'utf8');
        let data = yaml.load(content);
        if (!data) continue;

        for (let rootKey in data) {
            let items = data[rootKey];
            if (typeof items !== 'object') continue;

            for (let itemKey in items) {
                let itemData = items[itemKey];
                if (typeof itemData !== 'object') continue;

                let craftAmount = itemData['craft-amount'] || 1;
                let outputName = toTitleCase(itemKey);
                let outputVal = String(craftAmount) !== '1' ? outputName + "," + craftAmount : outputName;

                let crafting = itemData['crafting'] || {};
                let shaped = crafting['shaped'] || {};

                let frames = [];
                let shapeKeys = Object.keys(shaped).sort((a, b) => {
                    let anum = parseInt(a);
                    let bnum = parseInt(b);
                    if (!isNaN(anum) && !isNaN(bnum)) return anum - bnum;
                    return a.localeCompare(b);
                });

                for (let shapeKey of shapeKeys) {
                    let shapeData = shaped[shapeKey];
                    let inputs = shapeData['input'] || [];

                    let rows = [...inputs];
                    while (rows.length < 3) {
                        rows.push('v AIR 0 1..|v AIR 0 1..|v AIR 0 1..');
                    }

                    let frameQrs = [];
                    let rowLetters = ['A', 'B', 'C'];
                    
                    for (let rIdx = 0; rIdx < 3; rIdx++) {
                        let cols = rows[rIdx].split('|');
                        while (cols.length < 3) {
                            cols.push('v AIR 0 1..');
                        }
                        for (let cIdx = 0; cIdx < 3; cIdx++) {
                            let slotId = rowLetters[rIdx] + (cIdx + 1);
                            let itemName = parseSlotItem(cols[cIdx]);
                            if (itemName) {
                                frameQrs.push(slotId + ' "' + itemName + '"');
                            }
                        }
                    }

                    if (frameQrs.length > 0) {
                        frames.push(frameQrs.join(' '));
                    }
                }

                if (frames.length > 0) {
                    let qrsString = frames.join(' // ');
                    
                    let keyName = outputName;
                    let count = 2;
                    while (luaTable[keyName]) {
                        keyName = outputName + ' ' + count;
                        count++;
                    }

                    if (outputVal === outputName) {
                        luaTable[keyName] = [qrsString];
                    } else {
                        luaTable[keyName] = [qrsString, { Output: outputVal }];
                    }
                }
            }
        }
    } catch (e) {
        console.error("Error processing " + file, e);
    }
}

let luaLines = ["return {"];
for (let k in luaTable) {
    let v = luaTable[k];
    let escapedK = k.replace(/'/g, "\\'");
    if (v.length === 1) {
        luaLines.push("    ['" + escapedK + "'] = '" + v[0] + "',");
    } else {
        let outVal = String(v[1].Output).replace(/'/g, "\\'");
        luaLines.push("    ['" + escapedK + "'] = { '" + v[0] + "', Output = '" + outVal + "' },");
    }
}
luaLines.push("}");

fs.writeFileSync(outFile, luaLines.join('\n'), 'utf8');
console.log("Parsed " + Object.keys(luaTable).length + " items successfully.");

File: wiki/scratch/parse_crafting.py
import os
import glob
import yaml
import json

def to_title_case(s):
    return ' '.join(word.capitalize() for word in s.split('_'))

def parse_slot_item(item_str):
    item_str = item_str.strip()
    if item_str.startswith('v AIR') or not item_str:
        return None
    
    parts = item_str.split(' ')
    amount = "1"
    name = "Unknown"
    
    try:
        if item_str.startswith('v '):
            material = parts[1]
            name = to_title_case(material)
            # Find the element ending with '..'
            for p in parts:
                if '..' in p:
                    amount = p.split('.')[0]
        elif item_str.startswith('m '):
            if '-' in parts:
                idx = parts.index('-')
                id_part = parts[idx-1]
            else:
                id_part = parts[2]
            name = to_title_case(id_part)
            for p in parts:
                if '..' in p:
                    amount = p.split('.')[0]
    except Exception as e:
        print("Error parsing:", item_str, e)
        
    if amount.isdigit() and int(amount) > 1:
        return f"{name},{amount}"
    return name

lua_table = {}

yaml_files = glob.glob(r"e:\Project Wiki\crafting data\*.yml")
for yml_file in yaml_files:
    try:
        with open(yml_file, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)
    except Exception as e:
        print("YAML error in", yml_file, e)
        continue
    
    if not data: continue
        
    for root_key, items in data.items():
        if not isinstance(items, dict): continue
        for item_key, item_data in items.items():
            if not isinstance(item_data, dict): continue
            
            craft_amount = item_data.get('craft-amount', 1)
            output_name = to_title_case(item_key)
            if str(craft_amount) != '1':
                output_val = f"{output_name},{craft_amount}"
            else:
                output_val = output_name
                
            crafting = item_data.get('crafting', {})
            shaped = crafting.get('shaped', {})
            
            frames = []
            
            for shape_key in sorted(shaped.keys(), key=lambda x: int(x) if str(x).isdigit() else str(x)):
                shape_data = shaped[shape_key]
                inputs = shape_data.get('input', [])
                
                rows = inputs + ['v AIR 0 1..|v AIR 0 1..|v AIR 0 1..'] * (3 - len(inputs))
                
                frame_qrs = []
                row_letters = ['A', 'B', 'C']
                for r_idx in range(3):
                    cols = rows[r_idx].split('|')
                    cols = cols + ['v AIR 0 1..'] * (3 - len(cols))
                    for c_idx in range(3):
                        slot_id = f"{row_letters[r_idx]}{c_idx+1}"
                        item_name = parse_slot_item(cols[c_idx])
                        if item_name:
                            frame_qrs.append(f'{slot_id} "{item_name}"')
                
                if frame_qrs:
                    frames.append(" ".join(frame_qrs))
            
            if frames:
                qrs_string = " // ".join(frames)
                if output_val == output_name:
                    lua_table[output_name] = [qrs_string]
                else:
                    lua_table[output_name] = [qrs_string, {"Output": output_val}]

lua_lines = ["return {"]
for k, v in lua_table.items():
    if len(v) == 1:
        lua_lines.append(f"    ['{k}'] = '{v[0]}',")
    else:
        out_val = v[1]["Output"]
        lua_lines.append(f"    ['{k}'] = {{ '{v[0]}', Output = '{out_val}' }},")
lua_lines.append("}")

with open(r"e:\Project Wiki\wiki\Module\Crafting_SLASH_Data.md", "w", encoding='utf-8') as f:
    f.write("\n".join(lua_lines))
print(f"Parsed {len(lua_table)} items.")
